[#2] User Dto 구현 및 Swagger 작성

J-MU opened the pull request:
## 📢 관련 이슈
- [x] #2 

## 📃 작업 사항
- 유저 도메인의 DTO설계
- Swagger 작성

## 📃 #2를 먼저 진행한 이유
- BaseResponse쪽에서 고민이 되는 상황이라 우선 DTO설계부터 먼저 진행하였다.

## 📃 DTO 구조
유저 도메인의 주요 클래스들의 상속관계를 표현하여 첨부한다. 추상클래스는 이탤릭체로 표기하였으나 그림상에서 뚜렷하게 표시되지 않아 배경색을 회색으로 표시하였다.
![image](https://github.com/f-lab-edu/Prostargram/assets/46997074/b6d8c5be-bae8-4345-8ae2-4d44c4bb88a9)

- commentAuthor, PostAuthor : Post Domain에서 사용될 예정.
- GetFollowDto  : 팔로우/팔로잉 목록 화면에서 userId,userName,profileImg,departmentName가 표시되므로 UserInfo를 구현하는 방식으로 구현
- GetProfilePage : 프로필 페이지에서 관심사를 제외한 모든 프로필 정보를 포함하고 추가로 게시글 수, 팔로워 수, 팔로잉 수가 표시되므로 ProfileInfo를 구현하여 구현
- GetProfileUpdatePageDto : 프로필 수정 화면에 필요한 정보들을 받아 오기 위해 UpdatePage를 상속받는 방식으로 구현
- UpdateProfileDto : GetProfileUpdatePageDto와 같은 필드들을 가지지만 서로 다른 API이므로 따로 구현 클래스를 설계

## ✔️ CheckPoint
- DTO 상속 구조가 적절할까?
- 여러 Dto에서 공통으로 쓰일 수 있거나 추상 클래스들을 model package에 넣어놓았다. model이라는 명이 적절한 이름일까?
- model에 해당하는 클래스 이름들이 적절한지 고민된다. 특히 BasicUserInfo, UserInfo는 적절한 이름이 떠오르지 않는다.
- dto package구조를 request, response, model로 구분하였지만 다른사람들 패키지 구조를 보면 대부분 domain별로 구분하였다. 무엇이 더 괜찮은 구조일까?
- DTO를 서로 같은 정보를 갖더라도 API별로 구분하였다. 예를 들어 UpdateProfileDto와 GetProfileUpdatePageDto는 UpdatePage 추상클래스를 구현하였고 둘 다 추가 필드가 없지만 구분하였다. GetOptions, UpdateOptions도 서로 같은 정보를 갖지만 구분하였다. API별로 언제든 기획(기능)이 변경 될 수 있으므로 서로 다른 클래스를 갖는게 적절하지 않을까 하는 생각이였는데 이게 맞을지 모르겠다.
- getProfileInfo 메서드의 반환 값으로 ProfileInfo를 작성하니 Swagger상에서 example value가 모두 표시되지 않았다. type parameter에 따라 GetProfileDto와 GetProfileUpdatePageDto가 반환될 수 있어서 반환 타입을 저렇게 잡았는데 방법이 없을까?

## ✔️ 이슈2번 구현이 완료되지 않았는데 PR올리는 이유
- User Domain의 여러 클래스가 Feed Domain을 작성하고 있는 은비님한테 필요하여 로그인 도메인은 구현하지 않고 바로 올린다.
- 프로필 페이지에서 피드를 가져오는 API의 경우 Feed Domain의 클래스가 필요하므로 은비님 작업 후 구현할 예정이다.

J-MU committed:
[feat] profile관련 DTO 설계

J-MU committed:
[feat] get profile info API 설계

J-MU committed:
[feat] Get Follow Dto API 구현
- 공통으로 사용되는 유저정보를 UserBasicInfo 추상 클래스로 따로 빼서
  선언
- GetFollowDto를 위 클래스를 상속받도록 설계(팔로워/팔로잉을 모두 담을
  수 있는 클래스)

J-MU committed:
[feat] SGET Follows API Swagger작성

J-MU committed:
[feat] enum중 requestParam 따로 구분

J-MU committed:
[feat] GET options API DTO설계

J-MU committed:
[feat] 개인 설정 상태 확인하기 API Swagger 구현

J-MU committed:
[feat] DTO ì 상속 구조 수정

J-MU committed:
[feat] UpdateProfileRequestDTO 작성

J-MU committed:
updaㅕㅔUPDATE I Swagger 설êã계

J-MU committed:
Follow 추가 APDTO 설계

J-MU committed:
팔로우/팔로잉 추가 API Swagger작성

J-MU committed:
팔로우 삭제 DTO 작성

J-MU committed:
[feat] 팔로우 삭제 API Swagger 작성

J-MU committed:
Options DTO 수정/조회 분리

J-MU committed:
[feat] ì설정 수정 API Swagger 작성

J-MU committed:
[refactor] 불필요한 Reference Type 을 Primitive Type으로 변경

J-MU committed:
[refactor] package명 변경

J-MU committed:
[chore] 불필요한 import파일 삭제

J-MU committed:
[refactor] List 자료형은 접미-s로 끝나도록 통일

J-MU committed:
[refactor] FeedAuthor -> PostAuthor 클래스 ã명 변경

eunbileeme commented on src/main/java/flab/project/data/dto/common/CommonProfileInfo.java:
```diff
@@ -0,0 +1,15 @@
+package flab.project.data.dto.common;
```
common package와 response package를 구분한 이유가 있으실까요?

코드 상으로는 response에서 쓰이는 데이터(CommonProfileInfo, HashTag)가 common에 존재하고 이를 분리하여 두신 걸로 보이는데, 구분한 이유가 정확히 무엇일지 궁금합니다.

J-MU replied:
현재 Common은 패키지 명을 model로 수정한 상태입니다.
model에는 직접적인 DTO보다는 Request,Response DTO에 가져다가 쓰일 수 있는 클래스나 추상클래스가 속해있는 패키지입니다. common은 request, response 모두에서 쓰일 수 있다는 의미로 오해의 소지가 있어 model로 수정하였습니다.

eunbileeme commented on src/main/java/flab/project/data/dto/response/GetFollowDto.java:
```diff
@@ -1,10 +1,12 @@
 import lombok.Getter;
 
 @Getter
+@JsonPropertyOrder({"followId", "userId", "userName", "profileImgUrl", "departmentName" })
```
위 순으로 지정하신 특별한 이유가 있으실까요?

J-MU replied:
특별한 이유는 없고 팔로우를 가져오는 DTO다 보니 followId가 제일 먼저 나왔으면 좋겠고 그 이후로도 제가 느끼기에 젤 자연스러운 순서로 지정했습니다.

eunbileeme commented on src/main/java/flab/project/controller/UserController.java:
```diff
@@ -35,5 +39,22 @@ public ProfileInfo getProfileInfo(@PathVariable("userId") Long userId, @RequestP
+            @PathVariable("userId") Long userId,
+            @RequestParam("type") GetFollowsType type
+    ) {
+        return null;
```
이후에는 return 값을 test로 지정해놓으셨던데, 이유가 무엇인가요?

J-MU replied:
return이 test로 되어있는 메서드들은 성공 여부만 BaseResponse를 통해 보내줄 생각이라 저렇게 구현했습니다.

eunbileeme commented on src/main/java/flab/project/controller/UserController.java:
```diff
@@ -83,4 +85,19 @@ public String updateProfile(
     }
+
+    @Operation(
+        summary = "팔로우/팔로잉 추가 API"
```
팔로워/팔로잉을 말씀하시는 게 맞다면 팔로우를 **팔로워**로 수정하시면 좋을 것 같아요!

J-MU replied:
수정하겠습니다

eunbileeme commented on src/main/java/flab/project/data/dto/request/DeleteFollowDto.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.Getter;
+
+@Getter
+@Schema(description = "팔로워/팔로잉 삭제 DTO")
```
저희가 기획 때 팔로워는 삭제, 팔로잉은 취소로 표기하자고 했던 것으로 기억하는데 기능 상으로는 취소든 삭제든 Delete로 동일하므로 "Delete"로 표기하신 게 맞을까요?

eunbileeme replied:
추가로 팔로워/팔로잉 삭제는 "DeleteFollowDto" 하나의 파일로 구현하실 건지 궁금합니다.

J-MU replied:
1. 네 맞습니다.
2. Follower,Following 모두 Follow 테이블에 저장되어 있기 때문에 followId만 알고 있으면 간편하게 삭제할 수 있어서 둘 다DeleteFollowDto로 구현하려고 합니다.

eunbileeme commented on src/main/java/flab/project/data/enums/requestparam/GetProfileRequestType.java:
```diff
@@ -1,4 +1,4 @@
-package flab.project.data.enums;
+package flab.project.data.enums.requestparam;
```
기존에 UPDATE_PAGE_REQUEST와 함께 enum으로 작성하신 이유와 함께 분리하신 이유가 궁금해요.

J-MU replied:
GetFollowsType, GetProfileRequestType을 말씀하시는게 맞나요? 
```
@GetMapping(value = "/users/{userId}/follows")
    public List<GetFollowDto> getFollows(
            @PathVariable("userId") Long userId,
            @RequestParam("type") GetFollowsType type
    ) {
        return null;
        }
```
GetFollowsType Enum을 위와 같이 받으면 Swagger에서는 아래와 같이 나타납니다.

![image](https://github.com/f-lab-edu/Prostargram/assets/46997074/4e6cc303-9ee4-4e10-b0ad-ca5ce688c11c)

이것을 프론트에서 보면 저렇게 두 가지만 type으로 보낼 수 있구나라는 걸 알 수 있습니다.
만약, 무시하고 다른 값을 넣어서 요청을 보낸다면 Spring에서 Bad Request로 처리해줍니다.

![image](https://github.com/f-lab-edu/Prostargram/assets/46997074/43f8529f-1880-4f45-bc15-8b32ab118d2a)

정리하자면,
1. Enum에 존재하는 값만 허용되고 Enum에 없는 값이 넘어온다면 BadRequest로 처리된다. 단순 String으로 받는다면 저희가 직접 Validation해줘야합니다.
2. 프론트엔드에서 Swagger를 보고 어떤 값이 허용 되는지를 쉽게 알 수 있다.

이렇게 두가지 이유로 Enum을 사용했습니다.

eunbileeme replied:
네, 맞습니다. GetFollowsType, GetProfileRequestType를 두고 질문드린거였어요.
사진 첨부를 통한 자세한 설명 감사합니다.
그런데, "단순 String으로 받는다면 저희가 직접 Validation해줘야합니다."에 대해서 좀 더 자세히 기재해 주실 수 있으실까요?
github에 docs화 되면 두고두고 보고 참고하기에 좋을 것 같아요.

J-MU replied:
만약 String이라면 저희가 의도한 String값이 왔는지를 저희가 직접 체크해줘야합니다. Spring에서는 허용되는 문자열이 무엇인지 알 방법이 없으니까요. 반면 Enum으로 설정해놓으면 전달받은 값이 ENUM에 존재하는 값이 아닌경우 위의 PostMan에서 보듯이 BadRequest가 반환됩니다.

eunbileeme commented on src/main/java/flab/project/data/dto/request/UpdateOptionsDto.java:
```diff
@@ -2,6 +2,6 @@
 import flab.project.data.dto.common.Options;
 
-public class PatchOptionsDto extends Options {
+public class UpdateOptionsDto extends Options {
```
POST를 제외한 나머지 HTTP Method의 명명 규칙은 유지하는 것으로 알고 있었는데, Patch -> Update로 변경하신 이유가 무엇인가요?

J-MU replied:
이거 저희가 어디서 논의했는지 기억이 잘 안나는데 오늘 저녁 회의 때 이야기해보시죠

eunbileeme commented on src/main/java/flab/project/data/dto/common/BasicUserInfo.java:
```diff
@@ -6,7 +6,7 @@
 public abstract class BasicUserInfo {
     @Schema(example = "1")
-    private Long userId;
+    private long userId;
```
저도 primitive type으로 수정하되, 어제 얘기했던 것처럼 멘토님과 이야기를 나눠보는 것으로 알고 있겠습니다!

J-MU committed:
[chore] 팔로우 오타 팔로워로 수정

J-MU committed:
[refactor] DTO명명규칙 Code Convention에 따라 수정

f-lab-jason commented on src/main/java/flab/project/data/dto/model/BasicUserInfo.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.Getter;
+
+@Getter
+public abstract class BasicUserInfo {
```
UserInfo는 attribute가 다른 경우가 있나요? (abstract class가 굳이 필요한가해서 질문드려요)
유저의 타입(=권한)만 다를 것 같은데 혹시 제가 놓친게 있는지 궁금해요

f-lab-jason commented on src/main/java/flab/project/data/dto/model/HashTag.java:
```diff
@@ -0,0 +1,12 @@
+    @Schema(example = "1")
+    private long hashTagId;
+
+    @Schema(example = "#aws")
```
여기 db에 들어가는 데이터는 # 포함인가요?

J-MU replied:
네 #이 없는 데이터를 다룰 일이 없을거같아 #포함해서 삽입할 예정입니다.

f-lab-jason commented on src/main/java/flab/project/data/enums/requestparam/GetFollowsType.java:
```diff
@@ -0,0 +1,6 @@
+package flab.project.data.enums.requestparam;
+
+public enum GetFollowsType {
```
아래 GetProfileRequestType도 그렇고 이 enum의 value가 프론트에서 호출할때 붙혀야하는건가요? 그러면 enum name이 따로 있으면 좋을 것 같아서요.
호출시에 `.../follows?type=GET_FOLLOWERS_REQUEST` 보다는 `.../follows?type=follower`같은 형태로 보낼 수 있게 하는게 더 낫지 않을까 싶어서 문의합니다.

eunbileeme replied:
enum name을 추가하시면, 저도 참고하면서 이해해보겠습니다 민욱님 👀

J-MU committed:
[refactor] abstract class -> class로 변경
- BasicUserInfo에서 BasicUser로 클래스명 변경
- BasicUser, User, Profile을 abstract로 설정하였으나, 불필요한 조치로
  판단하고 일반 구현 클래스로 변경

J-MU committed:
[refactor] 기획 변경으로 인한 Profile DTO변경
- 프로필 페이지에는 관심사가 없었으나, 프로필 페이지에서 관심사가
  표시되도록 변경되었음.
- 수정 페이지와 프로필 페이지의 차이는 팔로잉/팔로워수/게시글 수 밖에
  없으므로 Profile 클래스와 GetProfilePage 클래스를 사용하는 방식으로
  변경
  UpdatePage와 그 자식 클래스들은 삭제

J-MU committed:
[refactor] Options 클래스는 추상클래스일 필요가 없으나 실수로 인해 추상클래스로 선언되어있었음. 수정

J-MU committed:
[refactor] package구조 변경
- 기존 request,response별로 dto내부에 패키지 구조를 나눴으나 실제로는
  request, response모두에서 쓰이는 클래스가 있는 등, 좋지 않은 구조로
  판단되어 수정

J-MU committed:
[refactor] Follow DTO 수정
- 기존에는 CreateFollowDto, DeleteFollowDto로 구분되어 있었음. 테이블에
  followId없이 fromUserId, toUserId를 복합키로 가져가기로 결정함에 따라
  CreateFollowDto, DeleteFollowDto의 정보가 같아짐.
- UpdateFollowDto로 하기에는 팔로우 요청이 Update가 아니라는 문제가
  발생. 이는 Like에서도 똑같이 발생할 것이라 예상됨.
- 위 배경에 따라 DTO명명규칙을 Update에서 Put,Patch로 수정하고 DTO수정

J-MU committed:
[refactor] Dto명명규칙 변경으로 인í한 dto명 변경

J-MU committed:
[refactor] GetFollowDto삭제
- UserInfo가 구현 클래스로 변경되었고, Follow 테이블에서 followId가
  삭제됨
- 위 배경에 따라 굳이 GetFollowDto를 따로 관리할 필요가 없어져 User
  클래스를 사용하기로 결정

J-MU committed:
[refactor] Controller 함수명 변경

J-MU committed:
[refactor] putFollows 요청을 post에서 put으로 변경

eunbileeme commented on src/main/java/flab/project/data/dto/model/BasicUser.java:
```diff
@@ -4,7 +4,7 @@
 
 @Getter
-public abstract class BasicUserInfo {
+public class BasicUser {
```
멘토님께서 말씀하신 것처럼 UserInfo(추후, User로 이름 변경된 파일)의 attribute(속성, 즉 필드)가 다른 경우가 없기 때문에 일반 클래스로 바꾸신 게 맞을까요?

J-MU replied:
1. abstract class로 선언했던 클래스들은 개발자가 해당 클래스를 생성하지 않기를 원해서 abstract클래스로 만들었으나 abstract class의 원래 용도인 기능의 설계도로서의 역할을 하는게 아니기 때문에 class로 변경했습니다.
2. class로 변경후 기존 추상클래스를 구현한 비어있는 클래스들의 경우 멘토님께서는 굳이 그럴 필요없이 기존 클래스를 활용하시기를 권장하셔서 제거했습니다.

eunbileeme commented on src/main/java/flab/project/controller/UserController.java:
```diff
@@ -1,11 +1,11 @@
-import flab.project.data.dto.request.DeleteFollowRequestDto;
-import flab.project.data.dto.request.UpdateOptionsRequestDto;
-import flab.project.data.dto.response.GetFollowResponseDto;
-import flab.project.data.dto.response.GetOptionsResponseDto;
```
저도 참고하여 request 및 response, 그리고 model 패키지를 삭제하고 바로 dto package 아래에 두었습니다.

eunbileeme commented on src/main/java/flab/project/controller/UserController.java:
"테이블에 followId없이 fromUserId, toUserId를 복합키로 가져가기로 결정함에 따라 CreateFollowDto, DeleteFollowDto의 정보가 같아짐."라는 말씀은 ERD를 보고 이해했어요.
따라서 정보가 같아졌기 때문에 DeleteFollowDto 파일만 지우신걸까요?

J-MU replied:
네 DeleteFollowDto는 기존에는 followId를 통해 삭제하도록 구현하였는데 그러지않고 통일시켰습니다.

eunbileeme commented on src/main/java/flab/project/controller/UserController.java:
```diff
@@ -110,9 +105,10 @@ public String updateProfile(
         }
     )
-    @PostMapping(value = "/users/{userId}/follows")
+    @PutMapping(value = "/users/{userId}/follows")
```
둘의 차이가 뭐지 싶어서 찾아보니, @PutMapping이 멱등성의 의미를 가지고 있는 기능이었군요. 👍🏻

J-MU committed:
[refactor] enum타입에 name추가
